File: 04-Timer引起的内存泄漏/TimerLeakOC/TimerLeakOC/TimerHolder/HYLTimerHolder.m
//
//  HYLTimerHolder.m
//  TimerLeakOC
//
//  Created by JF on 2017/9/25.
//  Copyright © 2017年 HYL. All rights reserved.
//

#import "HYLTimerHolder.h"

@interface HYLTimerHolder () {
    
    NSTimer *_timer;
    BOOL _repeats;
}

- (void)onTimer:(NSTimer *)timer;

@end

@implementation HYLTimerHolder

- (void)dealloc {
    
    [self stopTimer];
}

- (void)startTimer:(NSTimeInterval)seconds
          delegate:(id<HYLTimerHolderDelegate>)delegate
           repeats:(BOOL)repeats {
    
    _timerDelegate = delegate;
    _repeats = repeats;
    if (_timer) {
        
        [_timer invalidate];
        _timer = nil;
    }
    _timer = [NSTimer scheduledTimerWithTimeInterval:seconds
                                              target:self
                                            selector:@selector(onTimer:)
                                            userInfo:nil
                                             repeats:repeats];
}

- (void)stopTimer {
    
    [_timer invalidate];
    _timer = nil;
    _timerDelegate = nil;
}

- (void)onTimer: (NSTimer *)timer {
    
    if (!_repeats) {
        
        _timer = nil;
    }
    if (_timerDelegate && [_timerDelegate respondsToSelector:@selector(onHYLTimerFired:)]) {
        
        [_timerDelegate onHYLTimerFired:self];
    }
}

@end

File: 04-Timer引起的内存泄漏/TimerLeakSwift/TimerLeakSwift/TimerHolder/HYLTimerHolder.swift
//
//  HYLTimerHolder.swift
//  TimerLeakSwift
//
//  Created by JF on 2017/9/25.
//  Copyright © 2017年 HYL. All rights reserved.
//

/**
 功能为
 对于非repeats的Timer, 执行一次后自动释放Timer
 对于repeats的Timer, 需要持有HYLTimerHolder的对象在析构时调用[HYLTimerHolder stopTimer]
 */

import UIKit

protocol HYLTimerHolderDelegate: NSObjectProtocol {
    
    func onHYLTimerFired(_ timerHolder: HYLTimerHolder);
}

class HYLTimerHolder: NSObject {

    var timerHolderTag: Int
    
    weak var timerDelegate: HYLTimerHolderDelegate?
    
    private var _timer: Timer?
    
    private var _repeats: Bool
    
    override init() {
        
        timerHolderTag = 0
        _repeats = false
        
        super.init()
    }
    
    deinit {
        
        stopTimer()
    }
    
    func startTimer(_ seconds: TimeInterval, delegate: HYLTimerHolderDelegate, repeats: Bool) {
        
        timerDelegate = delegate
        _repeats = repeats
        if let _ = _timer {
            
            self._timer!.invalidate()
            self._timer = nil
        }
        _timer = Timer.scheduledTimer(timeInterval: seconds, target: self, selector: #selector(onTimer(_:)), userInfo: nil, repeats: true)
    }
    
    func stopTimer() {
        
        _timer?.invalidate()
        _timer = nil
        timerDelegate = nil
    }
    
    func onTimer(_ timer: Timer) {
        
        if (!_repeats) {
            
            _timer = nil;
        }
        
        timerDelegate?.onHYLTimerFired(self)
    }
}
